cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -976,6 +1044,37 @@ private void addPluginsToMap(NamespaceId namespace, Id.Artifact parentArtifactId
     return null;
   }
 
+  private void addPluginsInRangeToMap(NamespaceId namespace, List<Id.Artifact> parentArtifacts,
+                                      Map<byte[], byte[]> columns,
+                                      SortedMap<ArtifactDescriptor, PluginClass> plugins) {
+    for (Map.Entry<byte[], byte[]> column : columns.entrySet()) {
+      // column is the artifact namespace, name, and version. value is the serialized PluginData
+      ArtifactColumn artifactColumn = ArtifactColumn.parse(column.getKey());
+      Id.Namespace artifactNamespace = artifactColumn.artifactId.getNamespace();
+      // filter out plugins whose artifacts are not in the system namespace and not in this namespace
+      if (!Id.Namespace.SYSTEM.equals(artifactNamespace) && !artifactNamespace.equals(namespace.toId())) {
+        continue;
+      }
+      PluginData pluginData = GSON.fromJson(Bytes.toString(column.getValue()), PluginData.class);
+
+      ArtifactDescriptor latestParentArtifact = null;
+      // filter out plugins that don't extend this version of the parent artifact
+      for (Id.Artifact parentArtifactId : parentArtifacts) {
+        if (pluginData.usableBy.versionIsInRange(parentArtifactId.getVersion())) {
+          if (latestParentArtifact == null ||
+            parentArtifactId.getVersion().compareTo(latestParentArtifact.getArtifactId().getVersion()) > 0) {
+            latestParentArtifact = new ArtifactDescriptor(
+              artifactColumn.artifactId.toArtifactId(),
+              Locations.getLocationFromAbsolutePath(locationFactory, pluginData.getArtifactLocationPath()));
+          }
+        }
+        if (latestParentArtifact != null) {
+          plugins.put(latestParentArtifact, pluginData.pluginClass);
```

Comment:
the map key is not the parent artifact. It's the plugin artifact. None of the looping and latest parent logic is needed, it results in the same ArtifactDescriptor regardless of what the parent is.